oak/server/storage/storage_read_channel.h

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef OAK_SERVER_STORAGE_READ_CHANNEL_H_
+#define OAK_SERVER_STORAGE_READ_CHANNEL_H_
+
+#include "absl/types/span.h"
+#include "oak/proto/storage.grpc.pb.h"
+#include "oak/server/channel.h"
+#include "oak/server/storage/storage_manager.h"
+
+namespace oak {
+
+// A channel implementation that reads from a StorageService.
+class StorageReadChannel final : public ChannelHalf {
```

Comment:
Should this extend https://github.com/project-oak/oak/blob/7d75624bd36812d3f7c6bef09b70402e0354190e/oak/server/channel.h#L84 instead?

BTW, @daviddrysdale maybe `channel_halves_` should keep track of `MessageChannel` instance rather than `ChannelHalf`? I guess we expect all channels to provide async methods for reading / writing?

Reply:
At the moment we've got a single (per-Node) numbering space for channel handles, regardless of direction, so `channel_halves_` maps to a channel half rather than a channel.  (To put it another way: when we get as far as creating channels at runtime, the creator will get *two* numbers to refer to the new channel, one for read and one for write.)

More generally, I'm idly wondering if the whole storage channel thing could be replaced with a couple of `MessageChannel` instances that connect to some C++ code (i.e. a pseudo-node, to use the terminology from `manager.proto`).  

But I'm not yet ramped up enough on the storage design to know if that's a sensible or a silly idea&hellip;